index.ts

```diff
@@ -0,0 +1,552 @@
+  next(): {value: Item, done: boolean};
+}
+
+// moves.ts
+
+// TODO: consider moving some top-level booleans from Move into MoveFlags
+interface MoveFlags {
+  authentic?: boolean; // Ignores a target's substitute.
+  bite?: boolean; // Power is multiplied by 1.5 when used by a Pokemon with the Ability Strong Jaw.
+  bullet?: boolean; // Has no effect on Pokemon with the Ability Bulletproof.
+  charge?: boolean; // The user is unable to make a move between turns.
+  contact?: boolean; // Makes contact.
+  dance?: boolean; // When used by a Pokemon, other Pokemon with the Ability Dancer can attempt to execute the same move.
+  defrost?: boolean; // Thaws the user if executed successfully while the user is frozen.
+  distance?: boolean; // Can target a Pokemon positioned anywhere in a Triple Battle.
+  gravity?: boolean; // Prevented from being executed or selected during Gravity's effect.
+  heal?: boolean; // Prevented from being executed or selected during Heal Block's effect.
+  mirror?: boolean; // Can be copied by Mirror Move.
+  mystery?: boolean; // Unknown effect.
+  nonsky?: boolean; // Prevented from being executed or selected in a Sky Battle.
+  powder?: boolean; // Has no effect on Pokemon which are Grass-type, have the Ability Overcoat, or hold Safety Goggles.
+  protect?: boolean; // Blocked by Detect, Protect, Spiky Shield, and if not a Status move, King's Shield.
+  pulse?: boolean; // Power is multiplied by 1.5 when used by a Pokemon with the Ability Mega Launcher.
+  punch?: boolean; // Power is multiplied by 1.2 when used by a Pokemon with the Ability Iron Fist.
+  recharge?: boolean; // If this move is successful, the user must recharge on the following turn and cannot make a move.
+  reflectable?: boolean; // Bounced back to the original user by Magic Coat or the Ability Magic Bounce.
+  snatch?: boolean; // Can be stolen from the original user and instead used by another Pokemon using Snatch.
+  sound?: boolean; // Has no effect on Pokemon with the Ability Soundproof.
+}
+
+type MoveCategory: 'Physical' | 'Special' | 'Status';
+
+type MoveTarget =
+  // single-target
+  'normal'|'any'|'adjacentAlly'|'adjacentFoe'|'adjacentAllyOrSelf'|
+  // single-target, automatic
+  'self'|'randomNormal'|
+  // spread
+  'allAdjacent'|'allAdjacentFoes'|
+  // side and field
+  'allySide'|'foeSide'|'all';
+
+interface SelfEffect {
+  volatileStatus?: ID;
+  boosts?: Partial<BoostsTable>;
+}
+
+interface SecondaryEffect {
+  chance?: number;
+  boosts?: Partial<BoostsTable>;
+  self?: SelfEffect;
+  status?: StatusName;
+  volatileStatus?: ID;
+}
+
+type ContestType = 'Cool' | 'Beautiful' | 'Cute' | 'Clever' | 'Tough';
+
+interface CommonMove<TypeT, ItemT, MoveT> {
+  accuracy: number | true;
+  basePower: number;
+  flags: MoveFlags;
+  pp: number;
+  target: MoveTarget;
+  type: TypeT;
+
+  alwaysCrit?: boolean;
+  alwaysHit?: boolean;
+  basePowerModifier?: number;
+  boosts?: Partial<BoostsTable> | false
```

Comment:
I'd rather us consistently use `null` for emulating `Maybe<T>`

Reply:
On one hand, me too, on the other hand, this requires us to write out or backfill all the optional properties.

Reply:
My idea that I put as a comment at the top but didn't implement out of laziness was to make all the `JSONFoo` fields include `| null` via TS helper types and then they could be set to null to delete as necessary, but the rich API `Foo` wouldn't need to be aware of nulls. I guess I missed it here because `| false` was used instead of `| null`, but perhaps there's a reason for that?

Reply:
I looked, this is only false for gen 3 `charge`, and could easily just be `null` there like my plan. Removed `| false`.